File: backend/src/main/java/com/example/backendservice/features/collector/service/CollectorTaskService.java
package com.example.backendservice.features.collector.service;

import com.example.backendservice.features.collector.dto.*;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.time.Instant;
import java.util.UUID;

/**
 * Service interface for Collector task management
 */
public interface CollectorTaskService {

    /**
     * View tasks assigned to the collector
     * Default statuses: ASSIGNED, ON_THE_WAY
     */
    Page<CollectorTaskResponse> viewAssignedTasks(UUID collectorId, Pageable pageable);

    /**
     * Accept an assigned task
     * Transition: ASSIGNED → ON_THE_WAY
     */
    AcceptTaskResponse acceptTask(UUID taskId, UUID collectorId);

    /**
     * Update task status
     * Allowed transitions from ON_THE_WAY: COLLECTED, FAILED, CANCELLED
     */
    CollectorTaskResponse updateTaskStatus(UUID taskId, UUID collectorId, UpdateTaskStatusRequest request);

    /**
     * Upload proof image for completed task
     * Only allowed when status = COLLECTED
     */
    CollectorTaskResponse uploadProof(UUID taskId, UUID collectorId, UploadProofRequest request);

    /**
     * Get job history (completed, failed, cancelled jobs)
     */
    Page<JobHistoryResponse> getJobHistory(UUID collectorId, Instant from, Instant to, Pageable pageable);

    /**
     * Get performance summary metrics
     */
    PerformanceSummaryResponse getPerformanceSummary(UUID collectorId);
}

File: frontend/src/shared/services/api/collectorTaskService.ts
import apiClient from './client';

export interface CollectorTaskResponse {
  id: string;
  collectorId: string;
  reportId: string;
  enterpriseId: string;
  status: 'ASSIGNED' | 'ON_THE_WAY' | 'COLLECTED' | 'FAILED' | 'CANCELLED';
  note: string | null;
  collectorProofImageUrl: string | null;
  assignedAt: string;
  acceptedAt: string | null;
  onWayAt: string | null;
  collectedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface UpdateTaskStatusRequest {
  status: string;
  note?: string;
}

export interface UploadProofRequest {
  proofImageUrl: string;
  actualWeightKg?: number;
  note?: string;
}

export interface CollectorTaskPage {
  content: CollectorTaskResponse[];
  totalElements: number;
  totalPages: number;
  number: number;
  size: number;
}

export const collectorTaskService = {
  getAssignedTasks: (collectorId: string, page = 0, size = 50): Promise<CollectorTaskPage> =>
    apiClient
      .get<CollectorTaskPage>(`/collector/${collectorId}/tasks`, { params: { page, size } })
      .then(r => r.data),

  acceptTask: (collectorId: string, taskId: string): Promise<CollectorTaskResponse> =>
    apiClient
      .patch<CollectorTaskResponse>(`/collector/${collectorId}/tasks/${taskId}/accept`)
      .then(r => r.data),

  updateTaskStatus: (
    collectorId: string,
    taskId: string,
    body: UpdateTaskStatusRequest
  ): Promise<CollectorTaskResponse> =>
    apiClient
      .patch<CollectorTaskResponse>(`/collector/${collectorId}/tasks/${taskId}/status`, body)
      .then(r => r.data),

  uploadProof: (
    collectorId: string,
    taskId: string,
    body: UploadProofRequest
  ): Promise<CollectorTaskResponse> =>
    apiClient
      .post<CollectorTaskResponse>(`/collector/${collectorId}/tasks/${taskId}/proof`, body)
      .then(r => r.data),

  getJobHistory: (
    collectorId: string,
    params?: { from?: string; to?: string; page?: number; size?: number }
  ): Promise<CollectorTaskPage> =>
    apiClient
      .get<CollectorTaskPage>(`/collector/${collectorId}/history`, { params })
      .then(r => r.data),
};
